### PR TITLE
[eval] Consolidate Eval Functions for Step-Wise and Non-Step Wise

### PR DESCRIPTION
**Pull request by kyuds**
Previously in `evaluate.py`, we had separate functions for `evaluate` and `evaluate_step_wise`. This poses some problems:
1. Maintainability: there is a maintenance burden of two functions.
2. There are some inconsistencies in logic.

Apart from combining the logic, we make some fixes in this PR (for step-wise eval):
1. Pass in `step_wise=True` for `concatenate_generator_outputs`.
2. Pretty print examples like non-step wise counterpart.
4. Add `rollout_metrics` to `eval_metrics`.
5. Fix bug in `validate_generator_output` call where previous step-wise eval function took in `generator_input`, not `len(generator_input["prompts"])`, which was a typing mismatch
6. Fix incorrect counting of `num_turns_list` when `eval_n_samples_per_prompt` > 1.

---

> [!NOTE]
> **Medium Risk**
> Changes how eval metrics and logging are derived for step-wise trajectories and multi-sample prompts; incorrect row alignment would skew reported scores, though new tests target the main edge cases.
> 
> **Overview**
> **Merges** `evaluate` and `evaluate_step_wise` into a single `evaluate()` that branches on `cfg.generator.step_wise_trajectories`. The trainer and eval-only entrypoint no longer pick between two functions.
> 
> The unified path uses **`_EvalRows`**, **`_rows_for_output`**, and **`_scored_view`** to keep output rows aligned with prompts (including multi-step trajectories and multiple dataloader batches). **Metrics and trajectory logging** use one row per trajectory (last step only when step-wise); **eval dumps** still write every step.
> 
> **Step-wise fixes** bundled in the same refactor: pass `step_wise` into `concatenate_generator_outputs` and `validate_generator_output`, surface **rollout metrics** on eval, use **pretty_print_example** for debug samples, and compute **`num_turns_list`** per full `TrajectoryID` (so pass@n repetitions are not summed).
> 
> **Tests** in `tests/train/test_eval.py` cover last-step scoring, turn counts, dumps, rollout metrics, and two-batch alignment for both modes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit dfdb6a7f4d72e842cdb75810391c030be2a780fc. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: skyrl/train/evaluate.py [hunk 1]
```diff
@@ -37,6 +39,114 @@
...
+    """
+
+    env_classes: List[str] = field(default_factory=list)
+    env_extras: List[Dict[str, Any]] = field(default_factory=list)
+    uids: List[str] = field(default_factory=list)
+    prompts: List[Any] = field(default_factory=list)
+
+    @property
+    def data_sources(self) -> List[Optional[str]]:
+        return [env_extra.get("data_source") for env_extra in self.env_extras]
```
[the remark this answers is not in the document]
Comment (kyuds) [reply skyrl/train/evaluate.py:56]: fixed

File: skyrl/train/evaluate.py [hunk 2]
```diff
@@ -37,6 +39,114 @@
...
+        )
+    by_instance = {
+        traj_id.instance_id: (env_class, env_extra, prompt)
+        for traj_id, env_class, env_extra, prompt in zip(
+            generator_input["trajectory_ids"],
+            generator_input["env_classes"],
+            generator_input["env_extras"],
+            generator_input["prompts"],
+        )
+    }
```
[the remark this answers is not in the document]
Comment (kyuds) [reply skyrl/train/evaluate.py:103]: fixed

File: skyrl/train/evaluate.py [hunk 3]
```diff
@@ -37,6 +39,114 @@
...
+    ``num_turns`` to the trajectory logger (``None`` lets the logger derive turns from loss masks).
+    """
+    if not step_wise:
+        return concat_generator_outputs, rows, None
+
+    is_last_step = concat_generator_outputs["is_last_step"]
+    keep = [i for i, last in enumerate(is_last_step) if last]
+    view: Dict[str, list] = {}
+    for key, value in concat_generator_outputs.items():
+        if isinstance(value, list):
```
[the remark this answers is not in the document]
Comment (kyuds) [reply skyrl/train/evaluate.py:135]: this is not true. in `skyrl.train.generators.utils.concatenate_generator_outputs`, timing fields are still concatenated as-is (look at line 294 where `result` is first instantiated). The reduction that greptile points out happens in `rollout_metrics`, but this is a non-list item anyways so it doesn't matter.

[the remark this answers is not in the document]
Comment (kyuds) [reply skyrl/train/evaluate.py:135]: cc @CharlieFRuan for one more verification

File: skyrl/train/evaluate.py [hunk 4]
```diff
@@ -37,6 +39,114 @@
...
+            assert len(value) == len(
+                is_last_step
+            ), f"Length mismatch: {len(value)} != {len(is_last_step)} for key {key}"
+            view[key] = [value[i] for i in keep]
```
[the remark this answers is not in the document]
Comment (kyuds) [reply skyrl/train/evaluate.py:139]: look at response to greptile review, which pointed out the same thing.

File: skyrl/train/evaluate.py [hunk 5]
```diff
@@ -37,6 +39,114 @@
...
+    # TODO (kyuds): this count is likely wrong when `eval_n_samples_per_prompt > 1`. `rows.uids`
+    # holds `TrajectoryID.instance_id`, i.e. the *prompt* id, which every repetition of that prompt
+    # shares; counting by it merges repetitions, so if repetition 0 of a prompt took 3 steps and
+    # repetition 1 took 2, both of their last-step rows report 5 turns. Only the wandb trajectory
+    # table's turn column is affected. Keying the count and the lookup on
+    # `TrajectoryID.to_string()` (instance + repetition) would fix it; preserved as-is here because
+    # this consolidation must not change behaviour.
+    step_counts = Counter(rows.uids)  # counted BEFORE the filter, as today
+    scored_rows = rows.select(keep)
+    return view, scored_rows, [step_counts[uid] for uid in scored_rows.uids]
```
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl/train/evaluate.py]: @kyuds Let's just fix this rather than keeping the same behavior!

File: tests/train/test_eval.py [hunk 6]
```diff
@@ -111,3 +116,169 @@ async def test_evaluate_computes_expected_metrics(dummy_config, tmp_path):
...
+    return tokenizer
+
+
+@pytest.mark.asyncio
+async def test_evaluate_step_wise_scores_last_step_only(dummy_config, tmp_path):
+    cfg = _configure_eval(dummy_config, tmp_path, step_wise=True)
+    generator = DummyGenerator(_step_wise_output())
+
+    metrics = await evaluate(
+        eval_dataloader=DummyStatefulDataLoader([_PROMPTS_BATCH]),
```
Comment (CharlieFRuan) [inline tests/train/test_eval.py:206]: I think this only tests a single batch and hence doesn't test `_EvalRows.extend()` and `concatenate_generator_outputs`. Let's add a 2-batch test as well.

Let's also test when `eval_n_samples_per_prompt > 1` after we fix the current incorrect behavior.

**Review by CharlieFRuan — APPROVED** [review]
Mostly looks good, thank you! Let's fix the eval_n_samples_per_prompt > 1 case and slightly more tests; then we can merge!

File: skyrl/train/evaluate.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kyuds) [reply skyrl/train/evaluate.py]: fixed

File: tests/train/test_eval.py (hunk 6 above)
Comment (kyuds) [reply tests/train/test_eval.py:206]: added more tests :)

File: skyrl/train/evaluate.py [hunk 7]
```diff
@@ -37,6 +39,115 @@
...
+
+    is_last_step = concat_generator_outputs["is_last_step"]
+    keep = [i for i, last in enumerate(is_last_step) if last]
+    view: Dict[str, list] = {}
+    for key, value in concat_generator_outputs.items():
+        if isinstance(value, list):
+            assert len(value) == len(
+                is_last_step
+            ), f"Length mismatch: {len(value)} != {len(is_last_step)} for key {key}"
+            view[key] = [value[i] for i in keep]
```
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl/train/evaluate.py]: Wrong, `trajectory_generation_times` is only local variables in `concatenate_genreator_outputs()`. Re-check @gemini-code-assist

[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl/train/evaluate.py]: I think this was the behavior before as well

**Comment by CharlieFRuan** [comment]
Nice, thank you! @kyuds
